### DISCUSSION TITLE
Application window size issues.

### DISCUSSION THREAD

**Discussion by aleph1** [Kode/Kore#578] [General]
I am in the midst or porting my project from Kore to Kinc and am encountering an issue with the resolution of the application on Mac OS.

If I use the following code the window launches fullscreen :

```
kinc_window_options_t kinc_window_options;
kinc_window_options.width = 1440;
kinc_window_options.height = 900;
kinc_init("VieTest", 1440, 900, &kinc_window_options, NULL);
```

However, if I assign the width and height to variables and utilize them for kinc_window_options and kinc_init the application's window only occupies half of the width and height of the screen's resolution.

```
int width = 1440;
int height = 900;
kinc_window_options_t kinc_window_options;
kinc_window_options.width = width;
kinc_window_options.height = height;
kinc_init("VieTest", width, height, &kinc_window_options, NULL);
```

Obviously I am missing something here, but I am confused. Also, does Kinc include a way to get the resolution of a device's screen, or does that code have to be written on a per-platform basis?

**Comment by RobDangerous** [comment]
display.h includes the functions for querying the screen resolution. Kinc applications are dpi-aware nowadays so they report the real resolution on a high-dpi screen - might that be the problem here?

**Comment by aleph1** [comment]
If I modify the Kinc-Samples / Texture-Kinc repo (git clone --recursive https://github.com/Kinc-Samples/Texture-Kinc.git) and add the following lines to the start of the kickstart function I am still getting incorrect dimensions on a RMBP (kinc_log outputs 800 x 600 on a 1440 x 900 MBP display):

```
int kickstart(int argc, char **argv) {
    kinc_display_mode_t kinc_display_mode = kinc_display_current_mode(kinc_primary_display());
    kinc_log(KINC_LOG_LEVEL_INFO, "%i x %i", kinc_display_mode.width, kinc_display_mode.height);
    ...
}
```

**Comment by aleph1** [comment]
Also, if I try to include <Kore/Display.h> in the Sources/Shader.c file I get a compile error with Xcode which I wasn't getting when using Kore prior to Kinc. Is this the intended outcome or a bug with Kinc?
![error](https://user-images.githubusercontent.com/485722/98423160-cc803880-205b-11eb-966c-fdbcb69b489d.jpg)

**Comment by RobDangerous** [comment]
Looks like I still have to actually implement the display API for macOS.
That error though - that's because you're including a C++ header in a C file. Just rename Shader.c to Shader.cpp if you want to use the Kore-API.

**Comment by aleph1** [comment]
The following code seems to work for the display API for macOS in terms of getting width and height. If the display index that is passed to the function is out of bounds I am not sure whether you expect an error to be thrown, or whether to return a kinc_display_mode_t with width, height of 0.

```
kinc_display_mode_t kinc_display_current_mode(int display) {
    NSArray* screens = [NSScreen screens];
    NSScreen* screen = screens[display];
    NSRect screenRect = [screen frame];
    kinc_display_mode_t dm;
    dm.width = screenRect.size.width;
    dm.height = screenRect.size.height;
    dm.frequency = 60;
    dm.bits_per_pixel = 32;
    return dm;
}
```

Let me know if I should do a pull request with this code, or whether you want me to tackle the problem in a different way.

For refresh rate it looks like this has to be used: https://developer.apple.com/documentation/coregraphics/quartz_display_services, however NSScreen doesn't provide access to the same ID as [CGDirectDisplayID](https://developer.apple.com/documentation/coregraphics/cgdirectdisplayid) which would require iterating the displays in a different way than currently implemented.

**Comment by RobDangerous** [comment]
Yes, please send a pull request. I love pull requests.
